torch/csrc/jit/passes/alias_analysis.cpp

```diff
@@ -21,74 +23,329 @@ bool shouldAnnotate(const Value* v) {
 }
 } // namespace
 
-AliasDb::AliasDb(std::shared_ptr<Graph> graph) : graph_(std::move(graph)) {
-  analyze(graph_);
+// class AliasTracker
+//
+// This class tracks the "A points to B" graph for all values, as well as
+// wildcards and writes. It is used by AliasDb to provide a higher-level API.
+//
+// NOTE: this implementation is not very efficient; it's designed to be easy to
+// mutate as you modify the graph.
+class AliasTracker {
+ public:
+  // Returns true iff `v` is present in the alias set tracker.
+  bool contains(const Value* v) const {
+    return map_.count(v);
+  }
+
+  bool writesTo(Node* n, const Value* v) const {
+    if (isWildcard(v)) {
+      return wildcardWriters_.count(n);
+    }
 
-  // Build helper indices
-  // NOTE: that these assume that AliasDb is immutable once constructed.
-  // - Alias set -> value mapping
-  for (const auto& pr : valueToAlias_) {
-    const auto value = pr.first;
-    const auto& aliasInfo = pr.second;
-    // We don't support composite types yet
-    AT_ASSERT(aliasInfo.containedTypes().size() == 0);
-    for (const auto aliasSet : aliasInfo.sets()) {
-      aliasToValue_[aliasSet].insert(value);
+    if (!map_.count(v)) {
+      return false;
     }
+
+    return map_.at(v)->writers.count(n);
   }
-}
 
-bool AliasDb::hasWildcard(const Node* n) const {
-  return wildcardNodes_.count(n) != 0;
+  // Whether `a` *may* point to `b`
+  bool pointsTo(const Value* a, const Value* b) const {
+    if (!map_.count(a)) {
+      return false;
+    }
+    if (isWildcard(a) || isWildcard(b)) {
+      return true;
+    }
+
+    // BFS the subtree where the root is `a`s element and the branches are the
+    // `pointsTo` relationships.
+    const auto root = map_.at(a);
+    return root->bfs(
+        [&](const Element* el) { return el->value == b; },
+        BfsDirection::POINTS_TO,
+        /*shortCircuit=*/true);
+  }
+
+  // Make `v` point at `to`.
+  //
+  // Returns true if this actually changed the AliasTracker state (i.e. if
+  // we've never heard about this relationship before)
+  bool makeAlias(const Value* v, const Value* to) {
```

Comment:
The name of this function makes it sound symmetric, but its implementation in wildcard handling is not symmetric.

Reply:
I'll change it makePointerTo or something like that